ETL for city of chicago dataset

urvi-occ opened the pull request:
PR to add city of chicago dataset in Chicagoland Pandemic Response commons.  [Dataset link](https://data.cityofchicago.org/Health-Human-Services/COVID-19-Daily-Cases-Deaths-and-Hospitalizations/naz8-j4nc)

This dataset have one row dedicated for record for each day with `lab_report_date` and counts for covid cases, deaths and hospitalization data over age group, gender, race and ethnicity. New records are added each day in the dataset so the initial pass in ETL would check for last submission date and only fetch data from last submitted date till today. 

A new file has created inside `covid19-etl/etl/` named `cityofchicago.py` . This dataset has to make submissions to `summary_location` , `summary_clinical` and `summary_group_demographics` : 

- As the dataset is based in Chicago, only one `summary_location` would be needed in project. ETL would do an entry only if there are no location entries in project. 
- This dataset adds data each day so for each day one `summary_clinical` entry would be made with Total covid cases, deaths and hospitalization data 
- `summary_group_demographics` would have data per day and as per age group, gender, race and ethnicity

paulineribeyre commented on covid19-etl/etl/cityofchicago.py:
```diff
@@ -0,0 +1,410 @@
+                    submitter_value
+                ]
+
+        else:
+            return
```
is it expected that we'd get a header that doesn't match any of the previous `if` statements? if not, we should raise an error here instead of silently returning

urvi-occ replied:
I am adding code to raise Exception here.


```
        else:
            raise Exception("This process only give summary of covid data on either AgeGroup, Race or Gender.")
```

paulineribeyre commented on covid19-etl/etl/cityofchicago.py:
```diff
@@ -0,0 +1,410 @@
+
+        elif submitter_value in gender:
+            age_group_dict["Gender"] = gender[submitter_value]
+            if submitter_value == "unknown_gender":
+                summary_group_demographics_value_dict[
+                    "gender"
+                ] = "Unknown or Left Blank"
```
why don't you use the `gender` mapping for this?

urvi-occ replied:
for male and female both `submitter_id` and `gender` column value have same values according to the mapping but for "unknown_gender", the submitter id is suppose to have `unknown` and the `gender` column should have value as 
`Unknown or Left Blank`


|original value from dataset | submitter_id | gender |
|---------------------------|--------------|--------|
| male | male | male |
| female | female | female |
| unknown_gender | unknown | Unknown or Left Blank |


So I am using the mapping define to map from original value from dataset to the value in submitter id and because only unknown was different, I used an if condition. let me know what do you think about this

paulineribeyre replied:
makes sense

paulineribeyre commented on covid19-etl/etl/cityofchicago.py:
```diff
@@ -0,0 +1,410 @@
+            age_group_dict,
+        )
+
+        if (
+            summary_group_demographics_submitter_id
+            not in self.summary_group_demographics
+        ):
+            self.summary_group_demographics[
+                summary_group_demographics_submitter_id
+            ] = summary_group_demographics_value_dict
```
i don't understand this 🤔 you're calling `add_to_summary_group_demographics` for each column, and only storing `summary_group_demographics_value_dict` the 1st time. Doesn't that mean you're missing some values?

paulineribeyre replied:
oh i see, i think i missed [this](https://github.com/uc-cdis/covid19-tools/blob/5d113e05da7f37afab4cba977286f13098102e6b/covid19-etl/etl/cityofchicago.py#L234-L236) last time

urvi-occ replied:
Ya no worries :)

paulineribeyre commented on covid19-etl/etl/cityofchicago.py:
```diff
@@ -0,0 +1,410 @@
+        # To add `summary_group_demographics` data for each date in dataset with total of cases, deaths and hospitalization records
+
+        # `summary_group_demographics` have summary of covid data for either AgeGroup, Race, Gender
+        age_group_dict = {"AgeGroup": "None", "Race": "None", "Gender": "None"}
```
for `age_group_dict` and `summary_group_demographics_value_dict`: you don't need to initialize the values since you set them later. `age_group_dict = {}` would work, and if a key is missing, the ETL will fail to submit, which is a better outcome than silently submitting null data

paulineribeyre replied:
ok makes sense, it could be renamed though, it doesn't only contain age group data. maybe submitter_id_dict

and what about the default values in `summary_group_demographics_value_dict`? do we need them?

urvi-occ replied:
- Yes I can change it to `submitter_id_dict`
- About `summary_group_demographics_value_dict`,  the reason for having it to initialize the dict for `summary_group_demographics` so every time it would have same structure and we can change the needed values

paulineribeyre replied:
for `summary_group_demographics_value_dict` i'll go back to my initial comment then:

> `{}` would work, and if a key is missing, the ETL will fail to submit, which is a better outcome than silently submitting null data

urvi-occ replied:
I added those to add default value columns `age_group`, `race`, `gender` and `ethnicity` as null in `summary_group_demographics_value_dict` so I don't have to add those in each conditions.

If I can proceed with other key not mentioned in the dict, I can pass the null value which would give me a dict like this,

{'age_group': 'less than 17', 'count': 2, 'deaths': 0, 'hospitalizations': 0, 'summary_clinicals': [{'submitter_id': 'summary_clinical_us_il_chicago_2022-04-29'}]}

which is currently this

{'age_group': 'less than 17', 'race': '', 'gender': '', 'ethnicity': '', 'count': 2, 'deaths': 0, 'hospitalizations': 0, 'summary_clinicals': [{'submitter_id': 'summary_clinical_us_il_chicago_2022-04-29'}] }

paulineribeyre replied:
i think your first example would work?

urvi-occ replied:
As long as it works while submitting data, I can make the change

urvi-occ replied:
I added changes as new commit

paulineribeyre commented on covid19-etl/etl/cityofchicago.py:
```diff
@@ -0,0 +1,410 @@
+    def submit_metadata(self):
+        # Submits the data in `self.summary_locations`, `self.summary_clinicals` and `self.summary_group_demographic` to Sheepdog.
+        print("Submitting data...")
+        if len(self.summary_locations) > 0:
```
this `if` is not necessary, `for sl in self.summary_locations.values()` will just do nothing if the dict is empty

paulineribeyre commented on covid19-etl/etl/cityofchicago.py:
```diff
@@ -0,0 +1,410 @@
+CITYOFCHICAGO_CDH_URL = "https://data.cityofchicago.org/resource/naz8-j4nc.csv"
+
+
+class CITYOFCHICAGO(base.BaseETL):
```
i think a unit test with a fake input CSV file, that initializes a `CITYOFCHICAGO` instance, then calls `files_to_submissions()`, and then checks the contents of `self.summary_locations`, `self.summary_clinicals` and `self.summary_group_demographics` would be very good. Is that something you would have time to add?

urvi-occ replied:
I am not sure where and how to add the test. Can you give me some instructions and example for that?

paulineribeyre replied:
the unit tests are here https://github.com/uc-cdis/covid19-tools/tree/master/covid19-etl/tests
give me a few, i'll push something for you to get started

urvi-occ replied:
Sure. Thank you :) I should be creating a test case which should include dummy data submission (adding data into `summary_location`, `summary_clinical` and `summary_group_demographics`) and retrieval of it from guppy and verify?

paulineribeyre replied:
@urvi-occ you can cherry-pick this commit: `594739a9f8935babe1ed8e7a9ebb49bcce1f90ae` (https://github.com/uc-cdis/covid19-tools/commit/594739a9f8935babe1ed8e7a9ebb49bcce1f90ae)

all you have to do is update the assertions with the expected data

and add tests for edge cases if you can think of any

urvi-occ replied:
That make sense. Thank you. Is it okay if I directly contribute to `test/cityofchicago` branch or should I make copy of that and make a PR to `test/cityofchicago`  and then ask for a merge to master?

paulineribeyre replied:
`git cherry-pick 594739a9f8935babe1ed8e7a9ebb49bcce1f90ae`

paulineribeyre replied:
oh and the name of the test should be updated, i copy/pasted and forgot to change it

urvi-occ replied:
For sure, I would be working on that soon

urvi-occ replied:
@paulineribeyre 

> oh and the name of the test should be updated, i copy/pasted and forgot to change it

Can you help me locate this 😅

paulineribeyre replied:
the test name here: https://github.com/uc-cdis/covid19-tools/blob/8c8a8b300b1668d84ce5dae1f286f75e6f11fed7/covid19-etl/tests/test_cityofchicago.py#L39

urvi-occ replied:
Hey, I have added test cases and some changes in the input file to check with the missing values. Let me know if the test cases I added are okay or not

urvi-occ commented on covid19-etl/etl/cityofchicago.py:
```diff
@@ -0,0 +1,410 @@
+from etl import base
+import sys
+
+CITYOFCHICAGO_CDH_URL = "https://data.cityofchicago.org/resource/naz8-j4nc.csv"
```
The dataset have data till the previous day, for example today is 04-26-2022 so it would have data till 04-25-2022. about the data looking weird from 2022-04-21 , they are not putting hospitalization data for some days. Their website have put out the information below.

> Because of the nature of data reporting to CDPH, hospitalizations will be blank for recent dates They will fill in on later updates when the data are received, although, as for cases and deaths, may continue to be updated as further data are received. 

([source](https://data.cityofchicago.org/Health-Human-Services/COVID-19-Daily-Cases-Deaths-and-Hospitalizations/naz8-j4nc/))

As the csv file would still have the column, currently the ETL would bypass this and put the value as zero. But we should come up with a solution when the hospitalization data is available. May I get suggestions on this?

paulineribeyre replied:
Mm the easiest (but not smartest lol) solution would be to resubmit the whole dataset every time, instead of querying the `latest_submitted_date` 🤔 
It sounds like they expect all 3 columns could be updated.
Maybe we could choose an arbitrary period of time to update? For example the `start_date` could be `latest_submitted_date - 1-2 month`?

urvi-occ replied:
As the data is being updated everyday, ideal way that it should add data everyday. For this, We can do one thing, I can add a condition to check if hospitalization data is coming null or not in `parse_row` function. If its not null (meaning they started passing it again), I can raise exception and error it out. we can either wait for exception to occur and wait to get solution at that point or I can code right now that if this condition occur, I parse the file, check from which to which row the hospitalization records are empty and change `summary_group_demographics` record accordingly. Our main concern is to show `cases` and `deaths` data.

paulineribeyre replied:
I read `for cases and deaths, may continue to be updated` as "cases and deaths could be updated for past data". But i think you're right, the case and death data won't change once we submit it.

paulineribeyre replied:
> I parse the file, check from which to which row the hospitalization records are empty and change summary_group_demographics record accordingly

do you mean you would re-submit all the rows that have hospitalization data?

urvi-occ replied:
Can we update the current records? If we can we can do it otherwise we can skip it for now and I can take it internally and come up with an answer what can we do about it.

paulineribeyre replied:
yes we can, that was my suggestion `resubmit the whole dataset every time`

but i'll be a lot of data. if we can make it smarter it'd be better. we could try querying all the records that have no hospitalization data, and only update those. That's also a pretty big query :s

urvi-occ replied:
My approach for this would be, create a condition to check if date is greater than 21th April 2022 (the date where they stopped giving hospital data) and check if hospitalization column is empty or not, if not empty then take the records from file from 21th April 2022 to till date and make an api call to update hospitalization data on those records. The only query I would be making to sheepdog is to update those record.

paulineribeyre replied:
that's very manual 🤔 once the records are updated we have to remove the check. And when they stop adding hospitalization data again, we have to add a new date. I would say we can't be babysitting the ETL like that, it should be automatic

urvi-occ replied:
Yeah That make sense. what appoarch do you think we can use to make it smarter? This idea was just something which came up to me. I can take time into this to think about the suitable way to make this happen.

urvi-occ replied:
When the hospitalization data would be back, The datasource won't be updating the old columns (i.e. records from older dates) but only add the data in the new record (i.e. the current date) and ETL can work even if the record is present or empty. So I think we should be good with current process. Wydt?

paulineribeyre replied:
my suggestions from previous comments:

> the easiest (but not smartest lol) solution would be to resubmit the whole dataset every time

> we could try querying all the records that have no hospitalization data, and only update those

but now that i think about it, i think there's no way in Guppy to query data that does _not_ have a value

Something else you could do is use the `last_submission_identifier` to store which date the hospitalization data started being missing. That would work if all the missing values are in a row after a specific date, and not randomly distributed in the dataset

urvi-occ replied:
So you are saying, I should defined `last_submission_identifier` as a field in `summary_group_demogrpahics`

paulineribeyre replied:
`last_submission_identifier` is a property in the project node

> That would work if all the missing values are in a row after a specific date, and not randomly distributed in the dataset

is that ^ the case?

urvi-occ replied:
yes that's the case. its rows after specific date. Let me accommodate in the code

paulineribeyre replied:
ok then we could store in `last_submission_identifier` the number of the last row with hospitalization data

paulineribeyre replied:
@urvi-occ if you submit a record with a `submitter_id` that already exists, it will automatically do an update instead of a create. So in your case you would just need to process the row again and it should work

urvi-occ replied:
oh. That make sense. Thank you

urvi-occ replied:
I made a new commit where I am resubmitting missing data. so the data would be pulled from date `last_submission_identifier` till date. I also changed the date format and made it `%Y-%m-%d` . I'll be adding code for test in few days.

urvi-occ commented on covid19-etl/etl/cityofchicago.py:
```diff
@@ -0,0 +1,410 @@
+                self.convert_str_to_int(row[1]),
+                self.convert_str_to_int(row[2]),
+                self.convert_str_to_int(row[3]),
+                datetime.strptime(file_latest_date, "%Y-%m-%d"),
```
I did that because the original date is the formate `"%Y-%m-%dT%H:%M:%S.%f"`  the row in summary clinical should  have formate of `"%Y-%m-%dT%H:%M:%S"` and the format for summary clinical submitter id should only have `"%Y-%m-%d"`. 

So the first conversion converts from original to the row format (removing %f) and the second conversion on line 344 is for the last conversion (removing time)

Let me know what do you think?

urvi-occ replied:
Because of new commit, the line number changed for this but I would still like your review for this.

paulineribeyre replied:
oh i see, makes sense
but in that case i think you are losing the hour/min/sec/ms information by the time you add it the summary_clinical record 🤔

urvi-occ replied:
Yes. This is the norms which I was given for summary group demographics and it didn't had any entry of H:M:S

type | project_id | submitter_id | age_group | race | gender | ethnicity | count | deaths | ***hospitalizations*** | summary_clinicals.submitter_id
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
summary_group_demographics | open-cityofchicago | summary_group_demographic_us_il_cook_2021-03-14_less_than_20_None_None | less than 20 |   |   |   | 35 | 0 | 0 | summary_clinical_us_il_chicago_2021-03-14

paulineribeyre replied:
oh okay. i think it's weird to enforce `date` as `%Y-%m-%dT%H:%M:%S` if we are removing the H:M:S information. In that case Y-M-D is enough. Wdym "the norms which I was given"? did someone from my team say we should do it this way? just curious where this is coming from

urvi-occ replied:
Sean gave me this sample files on what this data should look like and I am following those.

paulineribeyre replied:
i don't see the `date` in the example row you shared

tbh i think it should either be "%Y-%m-%d" only, or include the exact time info we got from the source dataset

urvi-occ replied:
So The date format `%Y-%m-%dT%H:%M:%S` is used for `summary_clinical`



type | project_id | submitter_id | count | deaths | hospitaliIzedCumulative | date | summary_locations.submitter_id
-- | -- | -- | -- | -- | -- | -- | --
summary_clinical | open-cityofchicago | summary_clinical_us_il_chicago_2021-03-14 | 185 | 2 | 10 | 2022-03-14T00:00:00.000 | summary_location_us_il_chicago

The date formate `"%Y-%m-%d"` is used for `summary_group_demographics` as mentioned above.

So I am taking `%Y-%m-%dT%H:%M:%S.%f` from input ,
converting it to `"%Y-%m-%d"` for summary_clinical_submitter_id and `%Y-%m-%dT%H:%M:%S` for date column,
and using ``%Y-%m-%dT%H:%M:%S` to get `"%Y-%m-%d"` for `summary_group_demographics_submitter_id`

Let me know what formats I should be using and I can make changes accordingly.

urvi-occ replied:
Oh I just saw date format in summary_clinical have `%Y-%m-%dT%H:%M:%S.%f`. My bad. So I would make a commit to remove the conversion from `%Y-%m-%dT%H:%M:%S.%f` to `%Y-%m-%dT%H:%M:%S`

urvi-occ replied:
I made a new commit with date formats changed.

paulineribeyre replied:
i don't know where you are seeing that, in the datasets i checked in prod, the `date` is in format YYYY-MM-DD. Can we use that everywhere in a consistent way? there is no need to store the time for daily data

urvi-occ commented on covid19-etl/etl/cityofchicago.py:
```diff
@@ -0,0 +1,410 @@
+        Here we are ignoring records which doesn't have any lab report dates
+        """
+        if row[0] != "":
+            row_dict = {}
```
Removing it

urvi-occ commented on covid19-etl/etl/cityofchicago.py:
```diff
@@ -0,0 +1,410 @@
+            }
+
+    def add_summary_clinical(
+        self, cases_total, deaths_total, hospitalizations_total, today
```
changing this to `record_date`

urvi-occ commented on covid19-etl/etl/cityofchicago.py:
```diff
@@ -0,0 +1,410 @@
+#  This ETL is for city of chicago dataset for COVID-19 Daily Cases, Deaths, and Hospitalizations (CDH)
```
I made following changes according to the first review by @paulineribeyre 

- Added ETL in Readme
- Moved convert_str_to_int outside of the cityofchicago class
- Changed City from “chicago” to “Chicago”
- Removed `self.summary_location_submitter_id` and `self.headers` as class variables they could be used as internal variables
- Removed method `get_unified_date_format`
- `datetime.strptime("2020-03-01", “%Y-%m-%d”)` to `= datetime(2020, 3, 1)`
- In `add_summary_clinical` 
    - changed `today` to `record_date`
    - Added exception in last else statement 
- In `parse_row`
    - Removed `row` condition and `row_dict` variable
- In `submit_metadata`, removed if condition for `summary_locations`

urvi-occ commented on covid19-etl/etl/cityofchicago.py:
```diff
@@ -0,0 +1,410 @@
+            day = "0{}".format(day)
+        return "-".join((year, month, day))
+
+    def convert_str_to_int(self, string):
```
I moved the `convert_str_to_int` function outside the class, is it the correct place for it?

paulineribeyre replied:
yes that's what i meant! thanks 😄

paulineribeyre commented on README.md:
```diff
@@ -30,6 +30,7 @@
 | [COV-482][cov-482] | NCBI-MANIFEST | [bucket][ncbi-bucket] | scheduled|
 | [COV-465][cov-465] | NCBI | [bucket][ncbi-bucket] | scheduled|
 | [COV-532][cov-532] | COM-MOBILITY | [here](com-mobility) | scheduled|
+| [COV-1151][cov-1151] | CITYOFCHICAGO | [here](cityofchicago) | scheduled|
```
the link doesn't work, it's broken for the row above this one too, it should be `[here][...]` not `[here](...)`

urvi-occ replied:
I changed it and made a new commit

paulineribeyre commented on covid19-etl/etl/cityofchicago.py:
```diff
@@ -333,23 +338,25 @@ def parse_row(self, row):
-                datetime.strptime(file_latest_date, "%Y-%m-%d"),
+        if not row or not row[0]:
+            return
+        record_date = datetime.strptime(row[0], "%Y-%m-%dT%H:%M:%S.%f").strftime(
+            "%Y-%m-%d"
+        )
```
oh looks like you could just do `row[0].split("T")[0]` since you only need a string

paulineribeyre commented on covid19-etl/etl/cityofchicago.py:
```diff
@@ -0,0 +1,417 @@
+        according to row mapping in the dataset, in each row
+        column 0 would be lab report date
+        column 1,2 and 3 would be total number of cases, deaths and hospitalization per day which would be used in summary_clinical
+        column 3 to rest would be for summary_group_demographics for Age group, race , gender and ethincity
```
```suggestion
        column 4 to rest would be for summary_group_demographics for Age group, race , gender and ethincity
```

paulineribeyre commented on covid19-etl/etl/cityofchicago.py:
```diff
@@ -372,28 +378,29 @@ def files_to_submissions(self):
-                sl_record.update(sl)
-                self.metadata_helper.add_record_to_submit(sl_record)
-            self.metadata_helper.batch_submit_records()
+        print(self.summary_group_demographics)
```
is this a debug statement that should be removed?

urvi-occ replied:
Yes. Making a new commit with all above changes Incorporated.

paulineribeyre commented on covid19-etl/etl/cityofchicago.py:
```diff
@@ -354,36 +367,44 @@ def files_to_submissions(self):
         today = datetime.today().replace(hour=0, minute=0, second=0, microsecond=0)
 
-        if latest_submitted_date == today:
+        if latest_submitted_date != None and latest_submitted_date == today:
```
`if latest_submitted_date == today:` did the same thing as `if latest_submitted_date != None and latest_submitted_date == today:` 🤔

urvi-occ replied:
I thought I would require an explicit condition to check. Editing this

paulineribeyre commented on covid19-etl/etl/cityofchicago.py:
```diff
@@ -354,36 +367,44 @@ def files_to_submissions(self):
+        # Submits the data in `self.last_submission_identifier`, `self.summary_locations`, `self.summary_clinicals` and `self.summary_group_demographic` to Sheepdog.
         print("Submitting data...")
+
+        print("Updating last submission identifier date for project")
+        self.metadata_helper.update_last_submission(
+            self.last_submission_identifier.split("T")[0]
+        )
```
let's do this after the other submissions, not before: if the other submissions fail, next time the job runs it should start at the same `last_submission_identifier` as before

urvi-occ replied:
okay

paulineribeyre reviewed (commented):
looks great, thanks for working through it with me 🙂 we'll just wait for the test

urvi-occ commented on covid19-etl/tests/test_cityofchicago.py:
```diff
@@ -22,25 +27,51 @@ def close(self):
     # run the ETL
     etl = get_test_etl()
-    etl.files_to_submissions()
+    etl.get_summary_location(etl.summary_location_submitter_id)
```
because files_to_submissions() is pulling data from the original dataset URL. so to get data from the file, I had to change it

urvi-occ replied:
I need a little help in this, in original ETL code, `files_to_submissions()` function calls `parse_cityofchicago_file` which uses URL of original dataset to get data. I am not sure how should I implement the test case for this. can you suggest any flow for it?

urvi-occ replied:
No It didn't, It is taking data from original website as a code segment below from cityofchicago

urvi-occ replied:
Thank you so much for this. I would be adding changes for it in the new commit.. :)

urvi-occ commented on covid19-etl/tests/test_cityofchicago.py:
```diff
@@ -57,21 +57,21 @@ def test_cityofchicago():
+                "province_state": "IL",
+                "projects": [{"code": "cityofchicago"}],
+            }
         }
```
Do you mean about how this code block is dented? `python -m black` command edited in this manner.

urvi-occ replied:
ya I noticed it now. Editing it.

urvi-occ commented on covid19-etl/tests/test_data/test_cityofchicago_input.csv:
```diff
@@ -1,11 +1,11 @@
 "2022-04-22T00:00:00.000","683","0","10","152","126","148","98","63","48","34","14","0","392","289","2","135","68","98","236","48","98","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","2","1","2","1","0","1","2","1","0","8","2","0","2","1","3","3","1","0"
 "2022-04-21T00:00:00.000","679","0","18","113","168","160","80","69","50","23","16","0","357","317","5","115","53","116","251","45","99","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","4","4","0","2","2","3","3","0","12","6","0","0","0","9","9","0","0"
-"2022-04-20T00:00:00.000","717","0","11","97","190","177","100","75","36","26","16","0","417","294","6","113","67","90","293","33","121","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","2","2","0","0","2","2","3","0","7","4","0","0","1","5","5","0","0"
+,"717","0","11","97","190","177","100","75","36","26","16","0","417","294","6","113","67","90","293","33","121","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","2","2","0","0","2","2","3","0","7","4","0","0","1","5","5","0","0"
```
it doesn't error out it just skips that rows and don't submit that data. and if that's missing from original dataset than we can't do anything :| Should it be errored out?

paulineribeyre replied:
i'm not sure, i think it might be a problem, if the date is missing from the real dataset and we skip the row, and they fix their dataset later, we will probably never submit this row

is this something you have really seen in the real dataset?

urvi-occ replied:
So I have seen missing data for hospitalization but not for the other columns.

paulineribeyre replied:
i would say the ETL should not accept this

urvi-occ replied:
Now I get it. I would be adding changes for it in the new commit.

urvi-occ commented on covid19-etl/tests/test_data/test_cityofchicago_input.csv:
```diff
@@ -1,11 +1,11 @@
 "2022-04-25T00:00:00.000","922","0",,"214","188","198","108","98","61","31","24","0","512","402","8","151","78","133","339","63","158","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0",,,,,,,,,,,,,,,,,,
 "2022-04-24T00:00:00.000","400","0","7","75","64","99","63","48","29","14","8","0","229","169","2","71","27","47","177","23","55","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","1","0","1","0","1","4","0","4","3","0","0","1","2","4","0","0"
-"2022-04-23T00:00:00.000","407","0","1","66","68","94","58","48","41","23","9","0","215","189","3","87","34","54","148","24","60","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","1","0","0","0","0","0","1","0","1","0","0","0","0","0"
+"2022-04-23T00:00:00","407","0","1","66","68","94","58","48","41","23","9","0","215","189","3","87","34","54","148","24","60","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","1","0","0","0","0","0","1","0","1","0","0","0","0","0"
```
we are not using the millisecond values so I guess it is fine even if it is skipped. Wydt?

paulineribeyre replied:
oh i see, it does `.split("T")[0]` so it works 👍

urvi-occ commented on covid19-etl/tests/test_data/test_cityofchicago_input.csv:
```diff
@@ -1,11 +1,11 @@
 "lab_report_date","cases_total","deaths_total","hospitalizations_total","cases_age_0_17","cases_age_18_29","cases_age_30_39","cases_age_40_49","cases_age_50_59","cases_age_60_69","cases_age_70_79","cases_age_80_","cases_age_unknown","cases_female","cases_male","cases_unknown_gender","cases_latinx","cases_asian_non_latinx","cases_black_non_latinx","cases_white_non_latinx","cases_other_non_latinx", ...
 "2022-04-29T00:00:00.000","7","0",,"2","0","0","1","0","1","2","1","0","6","1","0","2","1","2","2","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0",,,,,,,,,,,,,,,,,,
-"2022-04-28T00:00:00.000","340","0",,"62","76","64","55","32","24","19","8","0","198","139","3","70","35","52","121","16","46","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0",,,,,,,,,,,,,,,,,,
+"2022-04-28T00:00:00.000",,"0",,"62","76","64","55","32","24","19","8","0","198","139","3","70","35","52","121","16","46","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0",,,,,,,,,,,,,,,,,,
```
So I removed some values in the file to make edge tests. Apart from hospitalization data, if any data is missing, it is replaced by zero. currently in the original dataset only hospitalization data is updated later after the day passed so we are using our logic to get data later but I haven't put condition for cases and deaths column.

paulineribeyre replied:
ok, i have the same concern here, i think we should error out if this happens 🤔 have you ever seen this case in the real dataset?

urvi-occ replied:
No. I haven't.

paulineribeyre replied:
i would say the ETL should not accept this either

paulineribeyre replied:
> or it should error out for missing cases and deaths values and use last_submission_identifier only for missing hospitalization records?

i would choose this so we don't over-engineer, since we don't expect them to update cases and deaths, but your call

urvi-occ replied:
I agree, I committed the code to raise exception if cases and deaths data is not found.

paulineribeyre commented on covid19-etl/etl/cityofchicago.py:
```diff
@@ -0,0 +1,430 @@
+        if len(current_summary_location) > 0:
+            return
+        else:
+            self.summary_locations[summary_location_submitter_id] = {
+                "country_region": self.country,
+                "county": self.county,
+                "province_state": self.state,
+                "projects": [{"code": self.project_code}],
+            }
```
@urvi-occ i tried to run the ETL in QA and got this error:
```
Submitting summary_location data
[...]
Exception: Unable to submit to Sheepdog: 400
{"code":400,"created_entity_count":0,"entities":[{"action":null,"errors":[{"keys":["submitter_id"],"message":"'submitter_id' is a required property","type":"MISSING_PROPERTY"}],"id":null,"type":"summary_location","unique_keys":[{"project_id":"open-cityofchicago","submitter_id":null}],"valid":false,"warnings":[]}],"entity_error_count":1,"message":"Transaction aborted due to 1 invalid entity.","success":false,"transaction_id":316584,"transactional_error_count":0,"transactional_errors":[],"updated_entity_count":0}
```

urvi-occ replied:
Looks like summary_location might need submitter_id defined in the dict, I made a new commit. Please let me know if it errors out after that.

paulineribeyre replied:
@urvi-occ 
```
Submitting data...
Submitting summary_location data
  Submitting 1 records in batches of 100
Submission progress: 1/1
Submitting summary_clinical data
  Submitting 804 records in batches of 100

Exception: Unable to submit to Sheepdog: 400
[...] "errors":[{"keys":[],"message":"Additional properties are not allowed ('hospitaliIzedCumulative' was unexpected)","type":"ERROR"}]
```

urvi-occ replied:
My bad. There was an extra I in the key `hospitalizedCumulative`. I would be removing it and making a new commit soon

paulineribeyre commented:
it still failed in QA:
```
Submitting summary_group_demographic data
Exception: Unable to submit to Sheepdog: 400
[...] "message":"'submitter_id' is a required property"
```
and the unit test failed:
```
        if not row or not row[0]:
>           raise Exception(
                "This row doesn't have date field present, ETL did not run successfully"
            )
E           Exception: This row doesn't have date field present, ETL did not run successfully
```

urvi-occ commented:
> it still failed in QA:
> 
> ```
> Submitting summary_group_demographic data
> Exception: Unable to submit to Sheepdog: 400
> [...] "message":"'submitter_id' is a required property"
> ```
> 
> and the unit test failed:
> 
> ```
>         if not row or not row[0]:
> >           raise Exception(
>                 "This row doesn't have date field present, ETL did not run successfully"
>             )
> E           Exception: This row doesn't have date field present, ETL did not run successfully
> ```

I added code to have `submitter_id` in `summary_group_demographic` and the error was happening because there was an extra line in csv input file. I removed extra white line and made a commit

paulineribeyre commented:
@urvi-occ You should be able to run the job in the OCC QA environment for testing, right? It still errors
```
"message":"'less than 17' is not one of ['less than 20', '20 to 29', '30 to 39', '40 to 49', '50 to 59', '60 to 69', '70 to 79', 'greater than 80', '12 to 15', '16 to 64', 'greater than 65', 'Unknown', None]"
```

urvi-occ commented:
Hey, so there is a gap here. The data is from `less than 17` and `18 to 29` and the columns asking are from `less than 20` and `20 to 29`. I changed the column value for `age_group` to match it

paulineribeyre commented on covid19-etl/etl/cityofchicago.py:
```diff
@@ -113,8 +113,8 @@ def add_to_summary_group_demographics(
-            "0_17": "less than 17",
-            "18_29": "18 to 29",
+            "0_17": "less than 20",
+            "18_29": "20 to 29",
```
that's not great though, ages from 18 to 20 are mapped to the wrong value. i think the dictionary should be updated

paulineribeyre reviewed (approved):
Ok it's working now. Any changes you want to make before we merge?

urvi-occ commented:
No its all good. Ready to merge
